design-proposals/klip-66-multi-param-udafs.md

```diff
@@ -0,0 +1,165 @@
+
+## What is in scope
+
+* UDAFs that accept variadic parameters: `example(col_int_1, col_int_2, col_int_3...) -> result`
+* UDAFs that accept a specific number of parameters, potentially of different types: `example(col_int, col_string) -> result`
+* Variadic TopK - a version of `TOPK` that returns the top K values of one column, along with the values of other columns from those records.
+* `correlation(double, double)` - a UDAF that computes the correlation coefficient of two columns.
+
+## What is not in scope
+
+* UDAFs that return multiple columns. 
+
+## Design
+
+### Variadic parameters
+
+We will add a new class called `MultipleArgs` for users to represent variadic arguments, which is a wrapper for `List`:
+
+```
+public class MultipleArgs<T> {
+
+  private final List<T> values;
+
+  public MultipleArgs(final List<T> values) {
+    this.values = values;
+  }
+
+  public List<T> getValues() {
+    return values;
+  }
+}
+```
+
+The `List` contains the values of the columns specified in the function call. An example usage is:
+```
+// Returns the sum of all the values from multiple colunms
+private static <Integer> Udaf<MultipleArgs<Integer>, Long, Long> MultiColumnSum() {
+  return new Udaf<MultipleArgs<Integer>, List<Integer>, Long>() {
+
+    @Override
+    public List<Integer> initialize() {
+      return 0;
+    }
+
+    @Override
+    public List<Integer> aggregate(MultipleArgs<Integer> current, Long aggregate) {
+      return aggregate + current.getValues().stream().reduce(0, Integer::sum);
+    }
+
+    @Override
+    public List<Integer> merge(Long aggOne, Long aggTwo) {
+      return aggOne + aggTwo;
+    }
+
+    @Override
+    public Long map(Long agg) {
+      return agg;
+    }
+  };
+}
+```
+
+To create a function that accepts variadic parameters of any type, the `VariadicArgs` type parameter can be set to `Object`.
+
+### Specific parameters
+
+In order to specify multiple parameters of different types, we will use the [javatuples](https://www.javatuples.org/) library.
+For example, a UDAF that requires one integer parameter and one string parameter can be implemented as follows:
+
+```
+private static <Integer> Udaf<Pair<Integer, String>, List<Integer>, Long> example() {
+  return new Udaf<Pair<Integer, String>, List<Integer>, Long>() {
+
+    @Override
+    public List<Integer> initialize() {
+      return null;
+    }
+
+    @Override
+    public List<Integer> aggregate(Pair<Integer, String> current, List<Integer> aggregate) {
+      return null;
+    }
+
+    @Override
+    public List<Integer> merge(List<Integer> aggOne, List<Integer> aggTwo) {
+      return null;
+    }
+
+    @Override
+    public Long map(List<Integer> agg) {
+      return null;
+    }
+  };
+}
+```
+
+[Javatuples](https://www.javatuples.org/) only go up to 10 parameters, but if a user wants to use more then they could use `VariadicArgs<Object>` and manually add type checking to the `aggregate` function.
```

Comment:
Javatuples looks neat.  

Did you see any alternative libraries?  There's likely nothing to the code, but since the last release was a decade ago, I have some pause.